seqr/utils/auth_middleware.py

```diff
@@ -0,0 +1,91 @@
+import abc
+
+from django.contrib.auth.models import User
+from django.core.exceptions import PermissionDenied
+from django.urls import resolve
+from django.utils.deprecation import MiddlewareMixin
+
+from google.oauth2 import id_token
+from google.auth.transport import requests
+
+from seqr.views.utils.permissions_utils import ServiceAccountAccess
+
+from settings import SOCIAL_AUTH_GOOGLE_OAUTH2_KEY
+
+
+class CheckServiceAccountAccessMiddleware(MiddlewareMixin):
+
+    def process_request(self, request):
+        func, _, _ = resolve(request.path)
+        request.service_account_access = isinstance(func, ServiceAccountAccess)
+
+
+class DisableCSRFServiceAccountAccessMiddleware(MiddlewareMixin):
+
+    def process_view(self, request, callback, callback_args, callback_kwargs):
+        assert hasattr(request, 'service_account_access'), (
+            'The seqr DisableCSRFServiceAccountAccess middleware requires '
+            'CheckServiceAccountAccessMiddleware middleware to be installed. '
+            'Edit your MIDDLEWARE setting to insert '
+            '"seqr.utils.auth_middleware.CheckServiceAccountAccessMiddleware" before '
+            '"seqr.utils.auth_middleware.DisableCSRFServiceAccountAccessMiddleware".'
+        )
+
+        if request.service_account_access:
+            # only exempt CSRF if it's a service account access route
+            callback.csrf_exempt = True
+
+
+class BearerAuth(MiddlewareMixin, abc.ABC):
+    def process_request(self, request):
+        assert hasattr(request, 'service_account_access'), (
+            'The seqr GoogleBearerAuth middleware requires '
```

Comment:
Will there eventually be sub-classes for other cloud providers? If so you could lift this error message into a static variable which other classes can override so that you don't have `GoogleBearerAuth` in your error message when you're using Azure for example.

```python
assert hasattr(request, 'service_account_access'), self.MISSING_MIDDLEWARE_ERROR
```

Alternatively if you are sure that all middleware will be defined in the same place you could use an f-string:

```python
f'"seqr.utils.auth_middleware.{self.__class__.__name__}".'
```

Reply:
Ooh nice catch! I've just replaced it with an fstring for now - if other people do use this change, I think there could be a way to get it using the `inspect` module.